design-proposals/klip-8-interactive-queries.md

```diff
@@ -0,0 +1,143 @@
+
+The remainder of this one pager will propose a high-level approach for incrementally evolving KSQL from a stream-processing system towards a streaming database, releasing the highest ROI capabilities for users first.
+
+## Goals
+
+* Make KSQL fundamentally more useful to application developers by exposing the data already stored within it to conventional `SELECT` queries.
+* Simplify the infrastructural requirements for interactive queries against materialized rows generated by KSQL.
+
+## Scope
+
+* **In**
+  * The high-level functional capabilities comprising interactive queries will be described.
+  * This high-level functionality will be roughly broken down into progressive tiers of completeness.
+  * The intended user experience of interactive queries will be described.
+* **Out**
+  * No specific engineering solutions to any of the capabilities discussed in this one pager will be proposed.
+  * Performance challenges of interactive queries will not be discussed.
+  * No proposals will be made regarding transactionality/consistency semantics around interactive queries.
+
+## Approach
+
+There is a wide spectrum of complexity between minimally useful row lookups and arbitrarily complex SQL queries over materialized state. Adding support for even the most trivial queries will still add significant value for KSQL users, so we should consider interactive queries an ongoing, incremental effort spanning numerous KSQL releases.
+
+Initial limitations around the kinds of interactive queries KSQL can support will also be augmented by the flexibility that users still have when defining materialized views with continuous queries. For example, even if aggregates are not initially supported for interactive queries, they can still be used in a materialized view’s definition. The same is true for joins.
+
+Here we roughly subdivide the spectrum of interactive query complexity into five progressive tiers, each tier building on the one before it. Each of these tiers can be released separately, although it may make sense to combine some of them.
+
+*Please note that some of the example queries below use mock functions for illustrative purposes.*
+
+### Tier 0 - Simple key-based lookups
+
+Support for key-based lookups is the bare minimum starting point for interactive queries that will still add meaningful utility. This essentially amounts to a cosmetic SQL layer above RocksDB’s key-value interface. Key-based lookups will allow users to look up a single row, given a row key:
+
+* WHERE clause must be strictly of the form: `WHERE ROWKEY = <rowkey>`
+* Each query can therefore only target a single shard.
+* Arbitrary expressions should be allowed.
+* Examples:
+  * `SELECT x, y, z FROM table WHERE ROWKEY = ‘rowkey’`
+  * `SELECT x + y FROM table WHERE ROWKEY = ‘rowkey’`
+
+### Tier 1 - Key-based range queries
```

Comment:
I *think* there is a tier 0.5 : time-based range queries on single key.

Where the table has a windowed key, we can support time based queries on a single key. This would hit a single KSQL server, though may cross multiple db shards on that machine.

e.g. 

`SELECT … FROM table WHERE ROWKEY =48 AND WINDOW_START>'2019-10-01 13:23:00:00' AND WINDOW_END<'2019-10-02 13:23:00:00'`

Reply:
@big-andy-coates can we currently specify a timestamp as a string like this without a conversion function/formatter? That would be really great if we could do this, as using integral timestamps + converter/formatter will not be ideal for such a common paradigm.

Maybe we should just include this explicitly as part of Tier 0? I think being able to specify time ranges for interactive queries against windowed tables is pretty important. Without that, windowed queries are not really usable with IQ.

Reply:
I don't see why not. It's just a bit of string parsing.  We'll have a fixed format we support, and can add functionality to allow this format to be changed later.